### PR TITLE
dump: Output escaped character values

### LINKED ISSUE #2398: String literal dump is raw
Dumped string literals keep their raw value

I tried this code:

```rust
const t: &str = "\tlol\nhello\"this is ' in string";
```

I expected to see this happen:

`cat gccrs.ast-pretty.dump`

```rust
const t : &str = "\tlol\nhello\"this is ' in string";
```

Instead, this happened:

`cat gccrs.ast-pretty.dump`

```rust
const t : &str = "	lol
hello"this is ' in string";
```

### Meta

- a06cec181de

### PR DESCRIPTION
**Pull request by P-E-P**
Output escaped character values instead of their raw values in string.

Fixes #2398

### CODE REVIEW

File: gcc/rust/lex/rust-token.cc [hunk 1]
```diff
@@ -180,13 +229,13 @@ Token::as_string () const
       switch (get_id ())
 	{
 	case STRING_LITERAL:
-	  return "\"" + get_str () + "\"";
+	  return "\"" + escape_special_chars (get_str (), true) + "\"";
```
Comment (philberty) [inline gcc/rust/lex/rust-token.cc:232]: tiny style thing, no need to do anything but if you have multiple flags passed to a function like this i would tend to write it as:

```escape_special_chars (get_str(), true /*in_string_flag*/)```

Comment (philberty) [reply gcc/rust/lex/rust-token.cc:232]: Just makes it a little more readable without having to look at the function declaration you're calling

**Review by philberty — APPROVED** [review]
LGTM

File: gcc/rust/lex/rust-token.cc [hunk 2]
```diff
@@ -172,6 +173,54 @@ Token::get_str () const
...
+	  else
+	    ss << "\\'";
+	  break;
+	case '"':
+	  if (in_string)
+	    ss << "\\\"";
+	  else
+	    ss << c;
+	  break;
+	// Should we also replace unicode escape or \x digits ?
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-token.cc:214]: @tamaroning what do you think?

**Review by CohenArthur — APPROVED** [review]
Looks good!

File: gcc/rust/lex/rust-token.cc (hunk 2 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-token.cc:214]: I think we should also convert characters back into unicode escapes and byte literal (\x).
If string literal contains other newlines than \n, dump file is broken. But it requires utf-8 decoding.

File: gcc/rust/lex/rust-token.cc (hunk 1 above)
Comment (P-E-P) [reply gcc/rust/lex/rust-token.cc:232]: Actually I don't know why I kept it like this, even better than a comment I would have used an enum.

Comment (P-E-P) [reply gcc/rust/lex/rust-token.cc]: Rebased and improved.

File: gcc/rust/lex/rust-token.cc [hunk 3]
```diff
@@ -172,6 +173,43 @@ Token::get_str () const
...
+{
+  std::stringstream ss;
+  decltype (matches)::const_iterator result;
+  for (char c : source)
+    {
+      // FIXME: Also replace escaped unicode values and \x digits
+      if ((result = matches.find (c)) != matches.end ())
+	{
+	  ss << result->second;
+	}
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-token.cc:199]:
```suggestion
	  ss << result->second;
```
nit

File: gcc/rust/lex/rust-token.cc [hunk 4]
```diff
@@ -172,6 +173,43 @@ Token::get_str () const
...
+
+const std::map<char, std::string> matches = {
+  {'\t', "\\t"}, {'\n', "\\n"},	 {'\r', "\\r"},
+  {'\0', "\\0"}, {'\\', "\\\\"}, {'\v', "\\v"},
+};
+
+std::string
+escape_special_chars (const std::string &source, Context ctx)
+{
+  std::stringstream ss;
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-token.cc:191]:
```suggestion
  std::stringstream stream;
```

very small nit, but I always hvae trouble understanding what `ss` refers to while `stream` is clearer IMO

**Review by CohenArthur — APPROVED** [review]
I like the enum in the anonymous namespace :+1: LGTM

File: gcc/rust/lex/rust-token.cc (hunk 4 above)
Comment (P-E-P) [reply gcc/rust/lex/rust-token.cc]: Fixed!

File: gcc/rust/lex/rust-token.cc (hunk 3 above)
Comment (P-E-P) [reply gcc/rust/lex/rust-token.cc]: Fixed!

File: gcc/rust/lex/rust-token.cc [hunk 5]
```diff
@@ -172,6 +173,41 @@ Token::get_str () const
...
+};
+
+std::string
+escape_special_chars (const std::string &source, Context ctx)
+{
+  std::stringstream stream;
+  decltype (matches)::const_iterator result;
+  for (char c : source)
+    {
+      // FIXME: Also replace escaped unicode values and \x digits
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-token.cc:195]: Put in the issue number? last nit and I'll merge this haha

**Review by CohenArthur — APPROVED** [review]

File: gcc/rust/lex/rust-token.cc (hunk 5 above)
Comment (P-E-P) [reply gcc/rust/lex/rust-token.cc]: Done

**Pull request merged**